### PR TITLE
Tag counters

### PR DESCRIPTION
**Pull request by Haarolean**

### CODE REVIEW

File: odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java [hunk 1]
```diff
@@ -19,8 +20,13 @@
     extends AbstractCRUDController<Tag, TagsResponse, TagFormData, TagFormData, TagService>
     implements TagApi {
 
-    public TagController(final TagService entityService) {
-        super(entityService);
+    private final TagService tagService;
+    private final TagMapper tagMapper;
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java:24]: Why do you need the mapper here?

File: odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java [hunk 2]
```diff
@@ -46,7 +52,9 @@ public TagController(final TagService entityService) {
         @Valid final String query,
         final ServerWebExchange exchange
     ) {
-        return list(page, size, query);
+        return tagService.listMostPopular(query, page, size)
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java:55]: tagService is accessible via this.entityService

File: odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepository.java [hunk 3]
```diff
@@ -9,6 +10,8 @@
 
     List<TagPojo> listByDataEntityId(final long dataEntityId);
 
+    List<TagDto> listMostPopular(String query, final int page, final int size);
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepository.java:13]: `String query` lacks final

File: odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepositoryImpl.java [hunk 4]
```diff
@@ -40,6 +45,22 @@ public TagRepositoryImpl(final DSLContext dslContext) {
...
     }
 
+    @Override
+    public List<TagDto> listMostPopular(final String query, final int page, final int size) {
+        return dslContext
+            .select(TAG.asterisk())
+            .select(DSL.count(TAG_TO_DATA_ENTITY.TAG_ID))
+            .from(TAG)
+            .join(TAG_TO_DATA_ENTITY).on(TAG.ID.eq(TAG_TO_DATA_ENTITY.TAG_ID))
+            .where(queryCondition(query))
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepositoryImpl.java:55]: `listCondition(query)`

It also adds soft delete filter

File: odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepositoryImpl.java [hunk 5]
```diff
@@ -73,4 +94,10 @@ public void createRelations(final long dataEntityId, final Collection<Long> tags
             throw new RuntimeException(e);
         }
     }
+
+    private List<Condition> queryCondition(final String query) {
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepositoryImpl.java:98]: Whole method can be removed

File: odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java [hunk 6]
```diff
@@ -6,14 +6,40 @@
...
+import reactor.core.publisher.Mono;
 
 @Service
 public class TagServiceImpl
     extends AbstractCRUDService<Tag, TagsResponse, TagFormData, TagFormData, TagPojo, TagMapper, TagRepository>
     implements TagService {
 
-    public TagServiceImpl(final TagMapper entityMapper, final TagRepository entityRepository) {
-        super(entityMapper, entityRepository);
+    private final TagRepository repository;
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java:19]: Accessible via entityRepository

File: odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java [hunk 7]
```diff
@@ -6,14 +6,40 @@
...
+    public TagServiceImpl(final TagMapper entityMapper, final TagRepository repository) {
+        super(entityMapper, repository);
+        this.repository = repository;
+    }
+
+    public Mono<TagsResponse> listMostPopular(final String query, final int page, final int size) {
+        return Mono
+            .fromCallable(() -> repository.listMostPopular(query, page, size))
+            .map(tags -> {
+                final List<Tag> items = tags.stream()
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java:30]: Whole mapping logic can be moved into the mapper

File: odd-platform-api/src/main/java/com/provectus/oddplatform/dto/TagDto.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package com.provectus.oddplatform.dto;
+
+import java.time.LocalDateTime;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+
+@Value
+@RequiredArgsConstructor
+public class TagDto {
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/dto/TagDto.java:9]: Why do you need the dto? Jooq's pojo seems to have all of the fields present already

**Review by DementevNikita — CHANGES_REQUESTED** [review]

File: odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java (hunk 1 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java:24]: Turns out I do not. Removed.

File: odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java (hunk 2 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/controller/TagController.java:55]: Oof, okay

File: odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepositoryImpl.java (hunk 4 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepositoryImpl.java:55]: Aight, thanks!

File: odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java (hunk 6 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java:19]: done

File: odd-platform-api/src/main/java/com/provectus/oddplatform/dto/TagDto.java (hunk 8 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/dto/TagDto.java:9]: Jooq's dto is missing the 'usedCount' field which is not a native one.

File: odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepository.java (hunk 3 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/repository/TagRepository.java:13]: Done. Also fixed the checkstyle

File: odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java (hunk 7 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/service/TagServiceImpl.java:27]: Done

File: odd-platform-api/src/main/java/com/provectus/oddplatform/mapper/TagMapperImpl.java [hunk 9]
```diff
@@ -52,4 +53,13 @@ public TagsResponse mapTags(final List<Tag> tags) {
             .items(tags)
             .pageInfo(pageInfo(tags.size()));
     }
+
+    public Tag mapTag(final TagDto dto) {
+        final Tag tag = new Tag();
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/com/provectus/oddplatform/mapper/TagMapperImpl.java:58]: Can be chained:
```
new TagPojo()
            .setName(form.getName())
            .setImportant(form.getImportant());
```

Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/mapper/TagMapperImpl.java:58]: `TagPojo` can be, but `Tag`'s setters return void.

Comment (DementevNikita) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/mapper/TagMapperImpl.java:58]: my bad, that'd be
```
        return new Tag()
            .id(pojo.getId())
            .name(pojo.getName())
            .important(pojo.getImportant());
```

**Review by DementevNikita — CHANGES_REQUESTED** [review]

File: odd-platform-api/src/main/java/com/provectus/oddplatform/mapper/TagMapperImpl.java (hunk 9 above)
Comment (Haarolean) [reply odd-platform-api/src/main/java/com/provectus/oddplatform/mapper/TagMapperImpl.java]: Done

**Pull request merged**